File: bindings/lua/rns/interfaces.lua
-- SPDX-License-Identifier: Apache-2.0
-- Copyright (c) 2024-2026 Quad4.io

local ffi_mod = require("rns.ffi")
local errors = require("rns.errors")
local ffi = ffi_mod.ffi
local lib = ffi_mod.lib

local function interfaces_list(node, capacity)
	capacity = capacity or 32
	if capacity <= 0 then
		errors.raise(errors.Error.INVALID_ARG)
	end
	local raw = ffi.new("rns_interface_entry[?]", capacity)
	local written = ffi.new("size_t[1]", 0)
	local code = lib.rns_interfaces(node:handle(), raw, capacity, written)
	if code ~= errors.Error.OK and code ~= errors.Error.TRUNCATED then
		errors.map_code(code)
	end
	local out = {}
	for i = 0, tonumber(written[0]) - 1 do
		local e = raw[i]
		out[#out + 1] = {
			name = ffi_mod.cstr(e.name, 96),
			type_name = ffi_mod.cstr(e.type_name, 32),
			online = e.online ~= 0,
			enabled = e.enabled ~= 0,
			rx_bytes = tonumber(e.rx_bytes),
			tx_bytes = tonumber(e.tx_bytes),
			rx_packets = tonumber(e.rx_packets),
			tx_packets = tonumber(e.tx_packets),
		}
	end
	return out
end

return {
	interfaces_list = interfaces_list,
}

File: bindings/java/src/io/quad4/rns/Interfaces.java
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2024-2026 Quad4.io

package io.quad4.rns;

import com.sun.jna.ptr.LongByReference;
import io.quad4.rns.ffi.RnsLibrary;
import java.util.ArrayList;
import java.util.List;

/** Interface snapshot helpers. */
public final class Interfaces {
    private Interfaces() {}

    public static final class Info {
        public final String name;
        public final String typeName;
        public final boolean online;
        public final boolean enabled;
        public final long rxBytes;
        public final long txBytes;
        public final long rxPackets;
        public final long txPackets;

        Info(
                String name,
                String typeName,
                boolean online,
                boolean enabled,
                long rxBytes,
                long txBytes,
                long rxPackets,
                long txPackets) {
            this.name = name;
            this.typeName = typeName;
            this.online = online;
            this.enabled = enabled;
            this.rxBytes = rxBytes;
            this.txBytes = txBytes;
            this.rxPackets = rxPackets;
            this.txPackets = txPackets;
        }
    }

    public static List<Info> list(Node node, int capacity) {
        if (capacity <= 0) {
            throw new RnsException(RnsException.INVALID_ARG);
        }
        RnsLibrary.RnsInterfaceEntry[] entries =
                (RnsLibrary.RnsInterfaceEntry[]) new RnsLibrary.RnsInterfaceEntry().toArray(capacity);
        LongByReference written = new LongByReference();
        int code = RnsLibrary.INSTANCE.rns_interfaces(node.handle(), entries, capacity, written);
        if (code != RnsLibrary.RNS_OK && code != RnsLibrary.RNS_ERR_TRUNCATED) {
            Rns.check(code);
        }
        List<Info> out = new ArrayList<>((int) written.getValue());
        for (int i = 0; i < written.getValue(); i++) {
            RnsLibrary.RnsInterfaceEntry e = entries[i];
            out.add(
                    new Info(
                            RnsLibrary.cstr(e.name),
                            RnsLibrary.cstr(e.type_name),
                            e.online != 0,
                            e.enabled != 0,
                            e.rx_bytes,
                            e.tx_bytes,
                            e.rx_packets,
                            e.tx_packets));
        }
        return out;
    }
}
